packages/react/src/createHooks.ts

```diff
@@ -0,0 +1,116 @@
+import { Query, Context, UpdateType, INode, LiveResult } from '@aphro/runtime-ts';
+import counter from '@strut/counter';
+import { useCallback, useMemo, useRef, useSyncExternalStore } from 'react';
+import { suspend } from 'suspend-react';
+
+const count = counter('model-infra/CreateHooks');
+
+export type ContextPromise = Promise<Context>;
+
+export type UseQueryOptions = {
+  // If you want to limit what sorts of updates will cause the live query to re-fire
+  on?: UpdateType;
+  // If you want to cache the results of the query.
+  // This'll allow components to unmount and remount without losing previously queried state.
+  // Should we just compute the key for them?
+  // We can given the declarative nature of the queries...
+  key: string;
+};
+
+type QueryReturnType<Q> = Q extends Query<infer M> ? M : any;
+
+export function createHooks(contextPromise: ContextPromise) {
+  // suspends the subtree using a hook that depends on context initialization
+  // until the context has been initialized.
+  function suspendOnContext() {
+    return suspend(() => {
+      return contextPromise;
```

Comment:
Curious how this works. Do future calls to suspend check that the provided promise is referentially equal to the prior promise and, if so, return the cached result synchronously?

Reply:
It's all explicit with the cache key tuple (next argument). In this case it will always cache hit because the key is static. Which I think is correct since this factory is generating hooks always linked to this context and that can't be changed.

Reply:
Ok, I just realized this would not be correct if different Contexts can exist, since they'd all cache together. Although that doesn't seem relevant to the library today... I think object equality can be used in cache keys, so I could maybe even pass the promise as the key and get the behavior you're talking about.

Reply:
Correct, different contexts are not relevant today.

The original idea behind different contexts came from a privacy perspective and the thought that you could instantiate multiple viewers in the same app. One viewer shouldn't share the cache with another viewer since the results of their privacy checks may not match.

Having multiple viewers in a request was a common pattern @ meta but probably actually a horrible pattern that created more problems than it solved. 

I'm going to look into removing the ability to have any viewer other than the currently logged in user. This might also mean we can just use a single context for the entire app. Any issues you see on your end with a single context?

Reply:
Nah, can't think of any potential use case for that.